README.md
# Bacon.Circuit

Bacon.js plugin for easily constructing stateful software components as a set of interconnected observables.

Adds `Circuit` and `Circuit.Field` types to the core `Bacon` library object. Exposes field factory methods at `Circuit.Field.stream(.…)` and `Circuit.Property.property(.…)`, also available at `Bacon.EventStream.field(.…)` and `Bacon.Property.field(.…)` respectively.


## Overview

Functional reactive programming (FRP) with [Bacon.js](https://baconjs.github.io) is extremely powerful that makes building inherently asynchronous and real-time software a breeze.

But as always with great power comes great responsibility and that is where Bacon.Circuit comes in. Consider it a circuit board of observable fields that collectively represent a piece of behavior and a corresponding state.

It offers a basic framework for constructing a stateful software component based on observables, which keeps you away from common FRP pitfalls _(hi there lazy evaluation)_ and makes it plain simple to integrate your component with other (possibly non-streaming) elements of your software by automatically generating a programming interface made up of conventional constructs such as getters, setters and methods.

Take this example of your typical data source module. A Node.js environment is assumed here, but it works in the browser in the same way.

```js
// mydatasource.js

var Bacon = require('baconjs');
require('bacon.circuit');

var request = require('superagent');
var ENDPOINT = 'http://httpbin.org/';

Bacon.Circuit.prototype.onEvent = function (name, observable, event) {
	console.log(name, '=', event.constructor.name, event.hasValue() && event.value());
};

// Note that in environments where a native ES6 `Promise` does not exist, a
// promise constructor needs to be defined in `Bacon.Circuit.prototype.
// promiseConstructor` to enable functions to return promises.

new Bacon.Circuit(module.exports, {
	
	config: {
		
		order: Bacon.Property.field.watch(function () {
			return Bacon.once("ASC");
		}),
		
		searchTerm: Bacon.Property.field.watch(function () {
			return Bacon.once("").debounce(100);
		})
		
	},
	
	issueQuery: Bacon.EventStream.field(function () {
		return Bacon.combineTemplate({
			order: this.config.order,
			q: this.config.searchTerm
		});
	}),
	
	value: Bacon.Property.field.digest(function () {
		return this.observeValue;
	}),
	
	observeValue: Bacon.Property.field.expose(function () {
		return Bacon.combineWith(
			function (params) {
				return params;
			}, this.issueQuery,
			Bacon.mergeAll(this.create, this.update, this.delete).startWith(true)
		).flatMapLatest(function (params) {
			return Bacon.fromNodeCallback(
				request.get(ENDPOINT + 'get').query(params), 'end'
			).map('.body');
		});
	}),
	
	create: Bacon.EventStream.field.method(function (data) {
		return Bacon.fromNodeCallback(
			request.post(ENDPOINT + 'post').send(data), 'end'
		).map('.body');
	}),
	
	update: Bacon.EventStream.field.method(function (item, data) {
		if (!item) return item;
		
		return Bacon.fromNodeCallback(
			request.put(ENDPOINT + 'put').query({ id: item.id }).send(data), 'end'
		).map('.body');
	}),
	
	delete: Bacon.EventStream.field.method(function (item) {
		if (!item) return item;
		
		return Bacon.fromNodeCallback(
			request.del(ENDPOINT + 'delete').query({ id: item.id }), 'end'
		).map('.body');
	})
	
});
```

Now we can use this module as follows:

```js
var ds = require('./mydatasource');

// Changing this property (and `config.searchTerm`) triggers that items will be
// refetched from server.
ds.config.order = "DESC";

// Function that returns a promise (same for `update` and `delete`).
ds.create({ id: 42, name: "Tim" }).then(function (response) {
	// ...
}, function (err) {
	// ...
});

// This value is assigned asynchronously, so not guaranteed to exist at this
// point.
if (ds.value === undefined)
	console.log("still loading items...");  

// This observable of type `Bacon.Property` has been exposed directly.
ds.observeValue.
	onValue(function (items) {
		// ...
	}).
	onError(function (err) {
		// ...
	});

// Note that `ds.issueQuery` does not exist.
```


## About

Bacon.Circuit was created by [Tim Molendijk](https://twitter.com/timmolendijk) as a building block of the [BangJS](http://bangjs.org) _Bacon-flavored Angular_ front-end application framework. It is currently in production use at [Nouncy](http://nouncy.com).

Contributing to Bacon.Circuit is encouraged. I would love to hear about your use case. Feel free to [email me](https://github.com/bangjs/bacon.circuit/blob/master/package.json#L22) or send in issues or pull requests.

Bacon.Circuit has been published under [MIT license](http://timmolendijk.mit-license.org/) and its [release history has been documented here](https://github.com/bangjs/bacon.circuit/blob/master/CHANGES.md).

CHANGES.md
### 0.2.4 :sparkles:
_Monday 5 October 2015_

* Expose circuit context on circuit instances.
* Fix bug that made properties on circuit context non-enumerable.


### 0.2.3 :sparkles:
_Sunday 6 September 2015_

* Fix bug in `watch` that resulted its merge callback from not being invoked with the right context (`this`).


### 0.2.2 :sparkles:
_Sunday 30 August 2015_

* Fix bug in `watch` that resulted in incorrect order of events that were sunk from merge function and duplicate event values when they were from merge function versus circuit watch.
* Minor tweak in NPM package configuration.


### 0.2.1 :sparkles:
_Thursday 9 July 2015_

* Fix bug in `watch` caused by superfluous asynchronicity (heh the irony).


### 0.2.0 :dizzy:
_Tuesday 7 July 2015_

* Introduce `sink` function as argument to field setup function. This essentially turns Bacon.Circuit into a framework for reactive programming, with *optional* functional operators. In other words; any field can now be implemented using nothing but subscribe (= read) and sink (= write) calls.
* Also pass observable instance as argument to field setup function, as a convenience to enable a field to subscribe to itself (useful for implementing side-effects) without having to parse its own name.
* Reorganization of CommonJS module: now exports `Bacon.Circuit` with field constructor at `Bacon.Circuit.Field` (and corresponding factory methods under `Bacon.EventStream.field` and `Bacon.Property.field`).
* Rename function as stream factory method from `function` to `method` to prevent reserved term issues in older ECMAScripts and to improve readability – no more `function(function () {…})`.
* Improve build process.
* Fix test that verifies that field observables never end.
* Minor facelift for CHANGES.


### 0.1.3 :sparkles:
_Monday 1 June 2015_

* Default promise constructor to make functions return promises out of the box on Node.js.
* Some improvements in README.


### 0.1.2 :sparkles:
_Sunday 31 May 2015_

* Fix many flawed tests.
* Improve overall Node.js support.
* Improve the test setup by making it support running them directly in Node.js.
* Improve the code sample in README and verify that it actually runs.
* Fix bug where initial stream values could get lost.
* Fix closure-in-loop bug and prevent others by replacing many occurences of `for..in` with safer alternative implementations.
* Fix bug where `stream.function` could miss values that were supposed to resolve its promise.
* Support watching properties that do not yet exist.


### 0.1.1 :sparkles:
_Wednesday 27 May 2015_

* Fix CommonJS support.
* Better code sample in README.


### 0.1.0 :zap:
_Wednesday 27 May 2015_

Initial release.


test/field.js
describe('Bacon.Circuit.Field', function () {
	
	it("instantiates field that delivers observable of specified type or `Bacon.Property` if not provided", function () {
		
		var field = new Bacon.Circuit.Field(_.noop, Bacon.EventStream);
		expect(field.observable()).to.be.instanceof(Bacon.EventStream);
	
		var field = new Bacon.Circuit.Field(_.noop, Bacon.Property);
		expect(field.observable()).to.be.instanceof(Bacon.Property);
	
		var field = new Bacon.Circuit.Field(_.noop);
		expect(field.observable()).to.be.instanceof(Bacon.Property);
		
	});
	
	it("delivers observable that never automatically ends", function (done) {
		
		var field = new Bacon.Circuit.Field(function (sink) {
			sink(1);
			_.delay(sink, 100, 3);
			return Bacon.constant(2);
		});

		field.observable().subscribe(_.after(3, function (event) {
			assert(!event.isEnd() && event.value() === 3);
			
			done();
		}));
		
		field.start();

	});
	
	it("can be started as soon as its observable has been subscribed to", function () {
		
		var field = new Bacon.Circuit.Field();
		
		expect(field).not.to.have.property('start');
		
		field.observable().subscribe(_.noop);
		
		expect(field).to.have.property('start').that.is.a('function');
		
	});
	
	it("executes setup on start and no sooner than that", function () {
		
		var onSetup = sinon.spy();
		
		var field = new Bacon.Circuit.Field(onSetup);
		expect(onSetup).to.have.not.been.called;
		
		field.observable().subscribe(_.noop);
		expect(onSetup).to.have.not.been.called;
		
		field.start();
		expect(onSetup).to.have.been.calledOnce;
		
	});
	
	it("calls setup with sink callback, observable and the parameters that were provided upon start", function (done) {
		
		var a = {}, b = 'b', c = true;
		
		var field = new Bacon.Circuit.Field(function (sink, me, name, circuit) {
			expect(this).to.equal(a);
			expect(sink).to.be.a('function');
			expect(me).to.equal(field.observable());
			expect(name).to.equal(b);
			expect(circuit).to.equal(c);
			
			done();
		});
		
		field.observable().subscribe(_.noop);
		
		field.start(a, b, c);
		
	});
	
	it("emits sunk values as events in generated observable", function (done) {
		
		var field = new Bacon.Circuit.Field(function (sink) {
			sink(1);
		});
		
		field.observable().onValue(function (value) {
			expect(value).to.equal(1);
			
			done();
		});
		
		field.start();
		
	});

	it("emits sunk values and returned stream values in order of appearance", function (done) {

		var field = new Bacon.Circuit.Field(function (sink) {
			sink(1);
			return Bacon.once(2);
		});

		field.observable().onValue(_.after(2, function (value) {
			expect(value).to.equal(2);

			done();
		}));

		field.start();

	});
	
	it("dismisses setup return value if not an observable of any sort", function (done) {
		
		var never = new Bacon.Circuit.Field(function () {
			return "ignore me";
		});
		
		Bacon.mergeAll(
			never.observable(),
			Bacon.once('first')
		).onValue(function (value) {
			expect(value).to.equal('first');
			
			done();
		});
		
		never.start();
		
	});
	
});

describe("Bacon.Circuit.Field.stream.expose", function () {
	
	it("assigns stream observable to circuit upon setup", function (done) {
		
		var field = Bacon.Circuit.Field.stream.expose(function (sink, me, name) {
			expect(circuit.set).
				to.have.been.calledOnce.
				to.have.been.calledWithExactly(name, field.observable());
			
			done();
		});
		
		var circuit = {
			set: sinon.spy()
		};
		field.observable().subscribe(_.noop);
		field.start({}, 'propName', circuit);
		
	});
	
});

describe("Bacon.Circuit.Field.stream.method", function () {
	
	it("assigns function to circuit upon setup", function (done) {
		
		var field = Bacon.Circuit.Field.stream.method();
		
		field.observable().subscribe(_.noop);
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				expect(fn).to.be.a.function;
				expect(fn()).to.be.undefined;
				
				done();
			}
		});
		
	});
	
	it("returns a promise with event value from a function call if a promise constructor is provided", function (done) {
		
		var field = Bacon.Circuit.Field.stream.method();
		
		field.observable().subscribe(_.noop);
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				assert(Q.isPromise(fn()));
				
				done();
			},
			promiseConstructor: Q.Promise
		});
		
	});
	
	it("captures every invocation of the function as a stream event", function (done) {
		
		var invoke;
		
		var field = Bacon.Circuit.Field.stream.method();
		
		field.observable().onValue(function (value) {
			expect(value).to.be.arguments;
			expect(_.toArray(value)).to.deep.equal([1, 2, 3]);
			
			done();
		});
		
		expect(invoke).to.be.undefined;
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				invoke = fn;
			}
		});
		
		expect(invoke).to.be.a.function;
		
		invoke(1, 2, 3);
		
	});
	
	it("can amend the result of the invocation using a `flatMapLatest` operation", function (done) {
		
		var onValue = sinon.spy();
		
		var field = Bacon.Circuit.Field.stream.method(function () {
			var args = _.toArray(arguments);
			return Bacon.once(args).startWith(args.map(function (n) {
				return n * n;
			})).map(function (numbers) {
				return numbers.reduce(function (sum, plus) {
					return sum + plus;
				}, 0);
			});
		});
		
		field.observable().onValue(function () {
			onValue.apply(this, arguments);
			
			if (onValue.calledTwice) {
				expect(onValue.firstCall).to.have.been.calledWithExactly(1*1 + 2*2 + 3*3);
				expect(onValue.secondCall).to.have.been.calledWithExactly(1 + 2 + 3);
				
				done();
			}
		});
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				fn(1, 2, 3);
			}
		});
		
	});
	
	it("resolves promise with first event after invocation", function (done) {
		
		var field = Bacon.Circuit.Field.stream.method(function (arg) {
			return Bacon.fromArray([arg, !arg]);
		});
		
		field.observable().subscribe(_.noop);
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				fn(true).done(function (value) {
					expect(value).to.be.true;
					
					done();
				});
			},
			promiseConstructor: Q.Promise
		});
		
	});
	
	it("drops pending events as soon as new (synchronous) invocation event comes in", function (done) {
		
		var onValue = sinon.spy();
		
		var field = Bacon.Circuit.Field.stream.method(function (value) {
			if (value === 1) return Bacon.later(50, 'drop me').startWith(value);
			return Bacon.once(value);
		});
		
		field.observable().onValue(onValue);
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				fn(1);
				fn(2);
				_.delay(fn, 100, 3);
			}
		});
		
		_.delay(function () {
			expect(onValue.firstCall).to.have.been.calledWithExactly(1);
			expect(onValue.secondCall).to.have.been.calledWithExactly(2);
			expect(onValue.thirdCall).to.have.been.calledWithExactly(3);
			
			done();
		}, 200);
		
	});

	it("drops pending events as soon as new (asynchronous) invocation event comes in", function (done) {
		
		var onValue = sinon.spy();
		
		var field = Bacon.Circuit.Field.stream.method(function (value) {
			if (value === 1) return Bacon.later(50, 'drop me').startWith(value);
			return Bacon.once(value);
		});
		
		field.observable().onValue(onValue);
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				fn(1);
				_.delay(fn, 0, 2);
				_.delay(fn, 100, 3);
			}
		});
		
		_.delay(function () {
			expect(onValue.firstCall).to.have.been.calledWithExactly(1);
			expect(onValue.secondCall).to.have.been.calledWithExactly(2);
			expect(onValue.thirdCall).to.have.been.calledWithExactly(3);
			
			done();
		}, 200);
		
	});
	
	it("drops pending events as soon as the actual invocation takes place, rather than the resulting event (if any)", function (done) {
		
		var onValue = sinon.spy();
		
		var field = Bacon.Circuit.Field.stream.method(function (value) {
			if (value === 1) return Bacon.later(200, 'drop me').startWith(value); 
			if (value === 2) return Bacon.never();
			return Bacon.once(value);
		});
		
		field.observable().onValue(onValue);
		
		field.start({}, 'propName', {
			set: function (name, fn) {
				fn(1);
				_.delay(fn, 50, 2);
				_.delay(fn, 400, 3);
			}
		});
		
		_.delay(function () {
			expect(onValue.firstCall).to.have.been.calledWithExactly(1);
			expect(onValue.secondCall).to.have.been.calledWithExactly(3);
			
			done();
		}, 500);
		
	});
	
});

describe("Bacon.Circuit.Field.property.digest", function () {
	
	it("assigns every value of property observable to circuit", function (done) {
		
		var onSet = sinon.spy();
		
		var field = Bacon.Circuit.Field.property.digest(function () {
			return Bacon.once(2).startWith(1);
		});
		
		field.observable().onValue(function (value) {
			if (value < 2) return;
			
			expect(onSet.firstCall).to.have.been.calledWithExactly('propName', 1);
			expect(onSet.secondCall).to.have.been.calledWithExactly('propName', 2);
			
			done();
		});
		
		field.start({}, 'propName', {
			set: onSet
		});

	});
	
	it("digests sunk values just as values from returned streams", function (done) {
		
		var field = Bacon.Circuit.Field.property.digest(function (sink) {
			sink(1);
		});
		
		field.observable().onValue(_.noop);
		
		field.start({}, 'propName', {
			set: function (key, value) {
				expect(value).to.equal(1);
				
				done();
			}
		});
		
	});
	
});

describe("Bacon.Circuit.Field.property.watch", function () {
	
	it("watches and reports changes of value on circuit", function (done) {
		
		var field = Bacon.Circuit.Field.property.watch();
		
		field.observable().onValue(function (value) {
			expect(value).to.equal(1);
			
			done();
		});
		
		field.start({}, 'propName', {
			watch: function (name, cb) {
				cb(1);
			},
			set: function () {}
		});
		
	});
	
	it("will merge the provided observable with (and before) the watch stream", function (done) {
		
		var onSet = sinon.spy();

		var field = Bacon.Circuit.Field.property.watch(function () {
			return Bacon.once(1);
		});
		
		field.observable().onValue(function (value) {
			if (value < 2) return;

			expect(onSet.firstCall).to.have.been.calledWithExactly('propName', 1);
			expect(onSet.secondCall).to.have.been.calledWithExactly('propName', 2);
			
			done();
		});
		
		field.start({}, 'propName', {
			watch: function (name, cb) {
				cb(2);
			},
			set: onSet
		});
		
	});

	it("calls merge with sink callback, observable and the parameters that were provided upon start", function (done) {
		
		var a = {}, b = 'b', c = {
			watch: _.noop,
			set: _.noop
		};
		
		var field = Bacon.Circuit.Field.property.watch(function (sink, me, name, circuit) {
			expect(this).to.equal(a);
			expect(sink).to.be.a('function');
			expect(me).to.equal(field.observable());
			expect(name).to.equal(b);
			expect(circuit).to.equal(c);
			
			done();
		});
		
		field.observable().subscribe(_.noop);
		
		field.start(a, b, c);
		
	});

	it("will merge sunk values with (and before) the watch stream", function (done) {
		
		var onSet = sinon.spy();

		var field = Bacon.Circuit.Field.property.watch(function (sink) {
			sink(1);
		});
		
		field.observable().onValue(function (value) {
			if (value < 2) return;

			expect(onSet.firstCall).to.have.been.calledWithExactly('propName', 1);
			expect(onSet.secondCall).to.have.been.calledWithExactly('propName', 2);
			
			done();
		});
		
		field.start({}, 'propName', {
			watch: function (name, cb) {
				cb(2);
			},
			set: onSet
		});
		
	});

	it("assigns every value of property observable to circuit", function (done) {

		var onSet = sinon.spy();
		
		var field = Bacon.Circuit.Field.property.watch(function () {
			return Bacon.once(2).startWith(1);
		});
		
		field.observable().onValue(function (value) {
			if (value < 3) return;

			expect(onSet.firstCall).to.have.been.calledWithExactly('propName', 1);
			expect(onSet.secondCall).to.have.been.calledWithExactly('propName', 2);
			expect(onSet.thirdCall).to.have.been.calledWithExactly('propName', 3);
			
			done();
		});
		
		field.start({}, 'propName', {
			watch: function (name, cb) {
				cb(3);
			},
			set: onSet
		});
		
	});
	
	it("will only issue events for actual value changes", function (done) {
		
		var onValue = sinon.spy();
		
		var o1 = {},
			o2 = {};
		
		var field = Bacon.Circuit.Field.property.watch(function (sink) {
			sink(undefined);
			return Bacon.once(1).startWith(undefined);
		});
		
		field.observable().onValue(function () {
			onValue.apply(this, arguments);
			
			if (onValue.callCount === 6) {
				expect(onValue.getCall(0)).to.have.been.calledWithExactly(undefined);
				expect(onValue.getCall(1)).to.have.been.calledWithExactly(1);
				expect(onValue.getCall(2)).to.have.been.calledWithExactly(2);
				expect(onValue.getCall(3)).to.have.been.calledWithExactly(undefined);
				expect(onValue.getCall(4)).to.have.been.calledWithExactly(o1);
				expect(onValue.getCall(5)).to.have.been.calledWithExactly(o2);
				
				done();
			}
		});
		
		field.start({}, 'propName', {
			watch: function (name, cb) {
				cb(1);
				cb(1);
				cb(2);
				cb(undefined);
				cb(o1);
				cb(o2);
			},
			set: function () {}
		});
		
	});

});

test/circuit.js
describe('Bacon.Circuit', function () {
	
	it("installs all provided fields with created context, sink function, its own observable, name and circuit instance", function () {
		
		var onSetup = sinon.spy(function () {
			expect(this.a).to.be.instanceof(Bacon.Observable);
			expect(this.b).to.be.instanceof(Bacon.Observable);
		});
		
		var a = new Bacon.Circuit.Field(onSetup);

		var circuit = new Bacon.Circuit({}, {
			a: a,
			b: new Bacon.Circuit.Field(_.noop)
		});
		
		expect(onSetup).to.have.been.calledWithExactly(sinon.match.func, a.observable(), 'a', circuit);
		
	});
	
	it("instantiates all observables before starting any of them", function (done) {
		
		var count = 0;
		
		new Bacon.Circuit({}, {
			a: new Bacon.Circuit.Field(function () {
				expect(this.b).to.be.instanceof(Bacon.Observable);
				
				if (++count === 2) done();
			}),
			b: new Bacon.Circuit.Field(function () {
				expect(this.a).to.be.instanceof(Bacon.Observable);
				
				if (++count === 2) done();
			})
		});
		
	});

	it("subscribes to all observables simultaneously, preserving all initial event stream events (via return)", function (done) {
		
		new Bacon.Circuit({}, {
			a: new Bacon.Circuit.Field(function () {
				return Bacon.once(1);
			}, Bacon.EventStream),
			b: new Bacon.Circuit.Field(function () {
				return this.a.doAction(function (value) {
					expect(value).to.equal(1);
					
					done();
				});
			}, Bacon.EventStream)
		});
		
	});
	
	it("subscribes to all observables simultaneously, preserving all initial event stream events (via sink)", function (done) {
		
		new Bacon.Circuit({}, {
			a: new Bacon.Circuit.Field(function (sink) {
				sink(1);
			}, Bacon.EventStream),
			b: new Bacon.Circuit.Field(function () {
				return this.a.doAction(function (value) {
					expect(value).to.equal(1);
					
					done();
				});
			}, Bacon.EventStream)
		});
		
	});
		
	it("takes in one to many field definition objects", function (done) {
		
		new Bacon.Circuit({}, {
			a: new Bacon.Circuit.Field(function () {
				expect(this.a).to.be.instanceof(Bacon.Observable);
				expect(this.b).to.be.instanceof(Bacon.Observable);
				
				done();
			})
		}, {
			b: new Bacon.Circuit.Field(_.noop)
		});
		
	});
	
	it("flattens the supplied list of field definition objects", function (done) {
		
		new Bacon.Circuit({}, [{
			a: new Bacon.Circuit.Field(function () {
				expect(this.a).to.be.instanceof(Bacon.Observable);
				expect(this.b).to.be.instanceof(Bacon.Observable);
				expect(this.c).to.be.instanceof(Bacon.Observable);
				
				done();
			})
		}, [{
			b: new Bacon.Circuit.Field(_.noop)
		}, {
			c: new Bacon.Circuit.Field(_.noop)
		}]]);
		
	});
	
	it("deeply merges multiple field definition objects", function (done) {
		
		new Bacon.Circuit({}, {
			a: new Bacon.Circuit.Field(function () {
				expect(this.nested.x).to.be.instanceof(Bacon.Observable);
				expect(this.nested.y).to.be.instanceof(Bacon.Observable);
				
				done();
			}),
			nested: {
				x: new Bacon.Circuit.Field(_.noop)
			}
		}, {
			nested: {
				y: new Bacon.Circuit.Field(_.noop)
			}
		});
		
	});
	
	it("has an `onEvent` method that is called for every event", function (done) {
		
		var field = new Bacon.Circuit.Field(function () {
			return Bacon.once('!');
		});
		var circuit = new Bacon.Circuit({}, {
			field: field
		});
		
		circuit.onEvent = function (name, observable, event) {
			expect(name).to.equal('field');
			expect(observable).to.equal(field.observable());
			assert(event.hasValue && event.value() === '!');
			
			done();
		};
		
	});
	
	it("has a `set` method that deep-assigns a value to provided interface object", function () {
		
		var circuit = new Bacon.Circuit({}, {});
		
		circuit.set('prop', true);
		expect(circuit.face.prop).to.be.true;
		
		circuit.set('nested.prop', 1);
		expect(circuit.face.nested.prop).to.equal(1);
		
		circuit.set('nested.prop', 2);
		expect(circuit.face.nested.prop).to.equal(2);
		
	});
	
	it("has a `watch` method that can register callbacks for changes on provided interface object", function () {
		
		var face = {};
		var circuit = new Bacon.Circuit(face, {});
		
		circuit.set('prop.nested', 1);
		
		var onWatch = sinon.spy();
		circuit.watch('prop.nested', onWatch);
		circuit.watch('prop.nested', onWatch);
		
		face.prop.nested = 2;
		
		expect(onWatch).to.have.been.calledTwice;
		expect(onWatch).to.have.always.been.calledWithExactly(2);
		
	});
	
	it("has a `watch` method that can register callbacks for changes on properties that only exist in the future", function (done) {
		
		var face = {};
		var circuit = new Bacon.Circuit(face, {});
		
		circuit.watch('prop', function (value) {
			expect(value).to.equal('first value');
			
			done();
		});
		
		face.prop = 'first value';
		
	});
	
});

test/conf/karma.js
// Karma configuration
// Generated on Mon May 25 2015 19:19:59 GMT+0200 (CEST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '../..',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'sinon-chai'],


    // list of files / patterns to load in the browser
    files: [
      'node_modules/lodash/index.js',
      'node_modules/q/q.js',
      'node_modules/baconjs/dist/Bacon.js',
      
      // TODO: Apply a preprocessor to these source files to parse them into
      // `src/build.template` and end up with a library that resembles
      // `dist/bacon.circuit.js` more closely (wrapped in closure and in strict
      // mode).
      'src/*.js',
      
      'test/*.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: [],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


src/field.js
function wrapSink(sink, ignoreEnd, emitAsync) {
	return function (value) {
		if (ignoreEnd && value instanceof Bacon.Event && value.isEnd())
			return;
		if (emitAsync)
			setTimeout(function () {
				sink(value);
			});
		else
			sink(value);
	};
}

function Field(setup, Type) {
	
	var observable = Bacon.fromBinder(function (sink) {
		sink = wrapSink(sink, true, true);
		
		this.start = function (context, name, circuit) {
			var result = setup.call(context, sink, this.observable(), name, circuit);
			
			if (result instanceof Bacon.Bus)
				result = result.toProperty();
			if (result instanceof Bacon.Property)
				result = result.toEventStream();
			if (result instanceof Bacon.EventStream)
				result.subscribe(sink);
			
			delete this.start;
			return this;
		};
		
		return function () {};
		
	}.bind(this));
	
	var doAction = function () {};
	this.doAction = function (fn) {
		doAction = fn;
	};
	
	observable = observable.doAction(function () {
		doAction.apply(this, arguments);
	});
	
	if (Type !== Bacon.EventStream)
		observable = observable.toProperty();
	
	this.observable = function () {
		return observable;
	};

}

Field.stream = function (setup) {
	return new Field(setup, Bacon.EventStream);
};

Field.property = function (setup) {
	return new Field(setup, Bacon.Property);
};

Field.stream.expose = Field.property.expose = function (setup) {
	return this(function (sink, me, name, circuit) {
		circuit.set(name, me);
		return setup.apply(this, arguments);
	});
};

Field.stream.method = function (flatMapLatest) {
	flatMapLatest = flatMapLatest || function () {
		return arguments;
	};
	return this(function (sink, me, name, circuit) {
		var context = this;
		return Bacon.fromBinder(function (latest) {
			circuit.set(name, function () {
				var stream = Bacon.once(arguments).flatMapLatest(function (args) {
					return flatMapLatest.apply(context, args);
				});
				
				if (!circuit.promiseConstructor) {
					latest(new Bacon.Next(stream));
					return;
				}
				
				return new circuit.promiseConstructor(function (resolve, reject) {
					latest(new Bacon.Next(stream.doAction(resolve).doError(reject)));
				});
			});
			return function () {};
		}).flatMapLatest(function (stream) {
			return stream;
		});
	});
};

Field.property.digest = function (setup) {
	var field = this(function (sink, me, name, circuit) {
		field.doAction(function (value) {
			circuit.set(name, value);
		});
		return setup.apply(this, arguments);
	});
	return field;
};

Field.property.watch = function (merge) {
	return this(function (sink, me, name, circuit) {
		return Bacon.fromBinder(function (send) {
			send = wrapSink(send, true);
			if (merge) {
				merge = merge.call(this, send, me, name, circuit);
				if (merge instanceof Bacon.Observable)
					merge.subscribe(send);
			}
			circuit.watch(name, send);
			return function () {};
		}.bind(this)).skipDuplicates().doAction(function (value) {
			circuit.set(name, value);
		});
	});
};

Bacon.Circuit.Field = Field;

Bacon.EventStream.field = Field.stream;
Bacon.Property.field = Field.property;

src/circuit.js
function Circuit(face) {
	// Calling without arguments can be done if we want to skip initialization
	// during inheritance.
	if (arguments.length === 0) return;
	
	var circuit = this;
	
	circuit.face = face;
	
	var fieldsObjs = [].slice.call(arguments, 1),
		fields = {};
	
	flattenArray(fieldsObjs).map(function (fieldsObj) {
		return unnestKeys(fieldsObj);
	}).forEach(function (fieldsObj) {
		Object.keys(fieldsObj).forEach(function (key) {
			if (fieldsObj[key] instanceof Bacon.Circuit.Field)
				fields[key] = fieldsObj[key];
		});
	});
	
	var keys = Object.keys(fields);

	circuit.context = {};
	
	keys.forEach(function (key) {
		// TODO: Making this an actual getter-setter is a bit pointless for
		// this scenario, but ah well doesn't really hurt either.
		setObjectProp(circuit.context, key, fields[key].observable());
	});
	
	keys.forEach(function (key) {
		fields[key].observable().subscribe(function (event) {
			circuit.onEvent(key, fields[key].observable(), event);
		});
	});

	keys.forEach(function (key) {
		fields[key].start(circuit.context, key, circuit);
	});
}

Circuit.prototype.set = function (key, value) {
	setObjectProp(this.face, key, value);
	return this;
};
Circuit.prototype.watch = function (key, cb) {
	setObjectProp(this.face, key);
	var leaf = findLeaf(this.face, key);
	var desc = Object.getOwnPropertyDescriptor(leaf.object, leaf.key);
	desc.set.listeners.push(cb);
	return this;
};
Circuit.prototype.onEvent = function () {};
Circuit.prototype.promiseConstructor = typeof Promise === 'function' && Promise;

function findLeaf(obj, path, create) {
	var keys = path.split('.');
	
	for (var i = 0; i < keys.length - 1; i++) {
		var key = keys[i];
		
		// TODO: Does this cover all scenarios in which we want to fail at
		// finding a leaf?
		if (!obj)
			return;
		
		if (create === true && !obj.hasOwnProperty(key))
			obj[key] = {};
		
		obj = obj[key];
	}
	return {
		object: obj,
		key: keys[i]
	};
}

function setObjectProp(obj, path, value) {
	var leaf = findLeaf(obj, path, true);
	
	if (!Object.getOwnPropertyDescriptor(leaf.object, leaf.key)) {
		
		var setter = function (v) {
			value = v;
			setter.listeners.forEach(function (listener) {
				listener(value);
			});
		};
		setter.listeners = [];
		
		Object.defineProperty(leaf.object, leaf.key, {
			configurable: true,
			enumerable: true,
			get: function () {
				return value;
			},
			set: setter
		});
		
	}
	
	if (arguments.length > 2)
		leaf.object[leaf.key] = value;
}

function unnestKeys(obj, path) {
	path = path || [];
	
	var flat = {};
	for (var key in obj) {
		if (!obj.hasOwnProperty(key)) continue;
		
		var keyPath = path.slice();
		keyPath.push(key);
		
		if (obj[key] instanceof Bacon.Circuit.Field) {
			flat[keyPath.join('.')] = obj[key];
			continue;
		}
		
		var nestedObj = unnestKeys(obj[key], keyPath);
		for (var nestedKey in nestedObj) {
			if (!nestedObj.hasOwnProperty(nestedKey)) continue;
			flat[nestedKey] = nestedObj[nestedKey];
		}
	}
	return flat;
}

function flattenArray(array) {
	var flat = [];
	array.forEach(function (item) {
		if (Array.isArray(item))
			flat.push.apply(flat, flattenArray(item));
		else
			flat.push(item);
	});
	return flat;
}

Bacon.Circuit = Circuit;

karma.conf.js
// Karma configuration
// Generated on Mon May 25 2015 19:19:59 GMT+0200 (CEST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'sinon-chai'],


    // list of files / patterns to load in the browser
    files: [
      'node_modules/lodash/index.js',
      'node_modules/q/q.js',
      'node_modules/baconjs/dist/Bacon.js',
      
      'src/*.js',
      
      'test/*.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: [],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};
